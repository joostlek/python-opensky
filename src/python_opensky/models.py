"""Asynchronous Python client for the OpenSky API."""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from .const import AircraftCategory, PositionSource
from .exceptions import OpenSkyCoordinateError
from .util import to_enum

if TYPE_CHECKING:
    from typing_extensions import Self


@dataclass(slots=True)
class StatesResponse:
    """Represents the states response."""

    states: list[StateVector]
    time: int

    @classmethod
    def from_api(cls, data: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            time=data["time"],
            states=[
                StateVector.from_api(vector_data) for vector_data in data["states"]
            ],
        )


@dataclass(slots=True)
# pylint: disable-next=too-many-instance-attributes
class StateVector:
    """Represents the state of a vehicle at a particular time.

    Attributes
    ----------
    icao24: ICAO24 address of the transmitter in hex string representation.
    callsign: Callsign of the vehicle.
    origin_country: Inferred through the ICAO24 address.
    time_position: Seconds since epoch of last position report. Can be None if there
     was no position report received by OpenSky within 15s before.
    last_contact: Seconds since epoch of last received message from this transponder.
    longitude: In ellipsoidal coordinates (WGS-84) and degrees.
    latitude: In ellipsoidal coordinates (WGS-84) and degrees.
    geo_altitude: Geometric altitude in meters.
    on_ground: True if aircraft is on ground (sends ADS-B surface position reports).
    velocity: Over ground in m/s.
    true_track: In decimal degrees (0 is north).
    vertical_rate: In m/s, incline is positive, decline negative.
    sensors: Serial numbers of sensors which received messages from the vehicle within
     the validity period of this state vector.
    barometric_altitude: Barometric altitude in meters.
    transponder_code: Transponder code aka Squawk.
    special_purpose_indicator: Special purpose indicator.
    position_source: Origin of this state's position.
    category: Aircraft category.
    """

    icao24: str
    callsign: str | None
    origin_country: str
    time_position: int | None
    last_contact: int
    longitude: float | None
    latitude: float | None
    geo_altitude: float | None
    on_ground: bool
    velocity: float | None
    true_track: float | None
    vertical_rate: float | None
    sensors: list[int] | None
    barometric_altitude: float | None
    transponder_code: str | None
    special_purpose_indicator: bool
    position_source: PositionSource
    category: AircraftCategory

    @classmethod
    def from_api(cls, data: dict[str, Any]) -> Self:
        """Initialize from the API."""
        return cls(
            icao24=data["icao24"],
            callsign=data.get("callsign"),
            origin_country=data["origin_country"],
            time_position=data.get("time_position"),
            last_contact=data["last_contact"],
            longitude=data.get("longitude"),
            latitude=data.get("latitude"),
            geo_altitude=data.get("geo_altitude"),
            on_ground=data["on_ground"],
            velocity=data.get("velocity"),
            true_track=data.get("true_track"),
            vertical_rate=data.get("vertical_rate"),
            sensors=data.get("sensors", []),
            barometric_altitude=data.get("baro_altitude"),
            transponder_code=data.get("squawk"),
            special_purpose_indicator=data["spi"],
            position_source=to_enum(
                PositionSource,
                data["position_source"],
                PositionSource.UNKNOWN,
            ),
            category=to_enum(
                AircraftCategory,
                data["category"],
                AircraftCategory.NO_INFORMATION,
            ),
        )


@dataclass(slots=True)
class BoundingBox:
    """Bounding box for retrieving state vectors."""

    min_latitude: float
    max_latitude: float
    min_longitude: float
    max_longitude: float

    def validate(self) -> None:
        """Validate if the latitude and longitude are correct."""
        self._check_latitude(self.min_latitude)
        self._check_latitude(self.max_latitude)
        self._check_longitude(self.min_longitude)
        self._check_longitude(self.max_longitude)

    @staticmethod
    def _check_latitude(degrees: float) -> None:
        if degrees < -90 or degrees > 90:
            msg = f"Invalid latitude {degrees}! Must be in [-90, 90]."
            raise OpenSkyCoordinateError(msg)

    @staticmethod
    def _check_longitude(degrees: float) -> None:
        if degrees < -180 or degrees > 180:
            msg = f"Invalid longitude {degrees}! Must be in [-180, 180]."
            raise OpenSkyCoordinateError(msg)
