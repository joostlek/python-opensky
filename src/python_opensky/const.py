"""Asynchronous Python client for the OpenSky API."""
import logging
from enum import Enum

LOGGER = logging.getLogger(__package__)


class PositionSource(int, Enum):
    """Enum holding the Position source."""

    UNKNOWN = -1
    ADSB = 0
    ASTERIX = 1
    MLAT = 2
    FLARM = 3


class AircraftCategory(int, Enum):
    """Enum holding the aircraft category."""

    NO_INFORMATION = 0
    NO_ADS_INFORMATION = 1
    LIGHT = 2
    SMALL = 3
    LARGE = 4
    HIGH_VORTEX_LARGE = 5
    HEAVY = 6
    HIGH_PERFORMANCE = 7
    ROTORCRAFT = 8
    GLIDER = 9
    LIGHTER_THAN_AIR = 10
    PARACHUTIST = 11
    ULTRALIGHT = 12
    RESERVED = 13
    UNMANNED_AERIAL_VEHICLE = 14
    SPACE_VEHICLE = 15
    EMERGENCY_VEHICLE = 16
    SERVICE_VEHICLE = 17
    POINT_OBSTACLE = 18
    CLUSTER_OBSTACLE = 19
    LINE_OBSTACLE = 20


MIN_LATITUDE = "lamin"
MAX_LATITUDE = "lamax"
MIN_LONGITUDE = "lomin"
MAX_LONGITUDE = "lomax"
