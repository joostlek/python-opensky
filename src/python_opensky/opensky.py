"""Asynchronous Python client for the OpenSky API."""
from __future__ import annotations

import asyncio
import math
import socket
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from importlib import metadata
from typing import TYPE_CHECKING, Any, cast

import async_timeout
from aiohttp import BasicAuth, ClientError, ClientResponseError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .const import MAX_LATITUDE, MAX_LONGITUDE, MIN_LATITUDE, MIN_LONGITUDE
from .exceptions import (
    OpenSkyConnectionError,
    OpenSkyError,
    OpenSkyUnauthenticatedError,
)
from .models import BoundingBox, StatesResponse

if TYPE_CHECKING:
    from typing_extensions import Self


@dataclass
class OpenSky:
    """Main class for handling connections with OpenSky."""

    session: ClientSession | None = None
    request_timeout: int = 10
    api_host: str = "opensky-network.org"
    opensky_credits: int = 400
    timezone = timezone.utc
    _close_session: bool = False
    _credit_usage: dict[datetime, int] = field(default_factory=dict)
    _auth: BasicAuth | None = None
    _contributing_user: bool = False

    async def authenticate(
        self,
        auth: BasicAuth,
        *,
        contributing_user: bool = False,
    ) -> None:
        """Authenticate the user."""
        self._auth = auth
        try:
            await self.get_states(bounding_box=BoundingBox(0.0, 0.0, 1.0, 1.0))
        except OpenSkyUnauthenticatedError as exc:
            self._auth = None
            raise OpenSkyUnauthenticatedError from exc
        self._contributing_user = contributing_user
        if contributing_user:
            self.opensky_credits = 8000
        else:
            self.opensky_credits = 4000

    @property
    def is_contributing_user(self) -> bool:
        """Return if the user is contributing to OpenSky."""
        return self._contributing_user

    @property
    def is_authenticated(self) -> bool:
        """Return if the user is correctly authenticated."""
        return self._auth is not None

    async def _request(
        self,
        uri: str,
        *,
        data: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Handle a request to OpenSky.

        A generic method for sending/handling HTTP requests done against
        OpenSky.

        Args:
        ----
            uri: the path to call.
            data: the query parameters to add.

        Returns:
        -------
            A Python dictionary (JSON decoded) with the response from
            the API.

        Raises:
        ------
            OpenSkyConnectionError: An error occurred while communicating with
                the OpenSky API.
            OpenSkyrror: Received an unexpected response from the OpenSky API.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host=self.api_host,
            port=443,
            path="/api/",
        ).joinpath(uri)

        headers = {
            "User-Agent": f"PythonOpenSky/{version}",
            "Accept": "application/json, text/plain, */*",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    METH_GET,
                    url.with_query(data),
                    auth=self._auth,
                    headers=headers,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the OpenSky API"
            raise OpenSkyConnectionError(msg) from exception
        except (
            ClientError,
            ClientResponseError,
            socket.gaierror,
        ) as exception:
            if isinstance(exception, ClientResponseError) and exception.status == 401:
                raise OpenSkyUnauthenticatedError from exception
            msg = "Error occurred while communicating with OpenSky API"
            raise OpenSkyConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")

        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from the OpenSky API"
            raise OpenSkyError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def get_states(
        self,
        bounding_box: BoundingBox | None = None,
    ) -> StatesResponse:
        """Retrieve state vectors for a given time."""
        credit_cost = 4
        params = {
            "time": 0,
            "extended": "true",
        }

        if bounding_box:
            bounding_box.validate()
            params[MIN_LATITUDE] = bounding_box.min_latitude
            params[MAX_LATITUDE] = bounding_box.max_latitude
            params[MIN_LONGITUDE] = bounding_box.min_longitude
            params[MAX_LONGITUDE] = bounding_box.max_longitude
            credit_cost = self.calculate_credit_costs(bounding_box)

        data = await self._request("states/all", data=params)

        if data["states"] is None:
            data["states"] = []
        else:
            data = {
                **data,
                "states": [self._convert_state(state) for state in data["states"]],
            }

        self._register_credit_usage(credit_cost)

        return StatesResponse.from_api(data)

    async def get_own_states(self, time: int = 0) -> StatesResponse:
        """Retrieve state vectors from your own sensors."""
        if not self._auth:
            raise OpenSkyUnauthenticatedError
        params = {
            "time": time,
        }

        data = await self._request("states/own", data=params)

        if data["states"] is None:
            data["states"] = []
        else:
            data = {
                **data,
                "states": [self._convert_state(state) for state in data["states"]],
            }

        return StatesResponse.from_api(data)

    @staticmethod
    def calculate_credit_costs(bounding_box: BoundingBox) -> int:
        """Calculate the amount of credits a request costs."""
        latitude_degrees = bounding_box.max_latitude - bounding_box.min_latitude
        longitude_degrees = bounding_box.max_longitude - bounding_box.min_longitude
        area = latitude_degrees * longitude_degrees
        if area < 25:
            return 1
        if area < 100:
            return 2
        if area < 400:
            return 3
        return 4

    def _register_credit_usage(self, opensky_credits: int) -> None:
        self._credit_usage[datetime.now(self.timezone)] = opensky_credits

    def remaining_credits(self) -> int:
        """Calculate the remaining opensky credits."""
        now = datetime.now(self.timezone)
        used_credits = sum(
            v
            for k, v in self._credit_usage.items()
            if now - timedelta(hours=24) <= k <= now
        )
        return self.opensky_credits - used_credits

    @staticmethod
    def _convert_state(state: list[Any]) -> dict[str, Any]:
        keys = [
            "icao24",
            "callsign",
            "origin_country",
            "time_position",
            "last_contact",
            "longitude",
            "latitude",
            "baro_altitude",
            "on_ground",
            "velocity",
            "true_track",
            "vertical_rate",
            "sensors",
            "geo_altitude",
            "squawk",
            "spi",
            "position_source",
            "category",
        ]

        return dict(zip(keys, state, strict=True))

    @staticmethod
    def get_bounding_box(
        latitude: float,
        longitude: float,
        radius: float,
    ) -> BoundingBox:
        """Get bounding box from radius and a point."""
        half_side_in_km = abs(radius) / 1000

        lat = math.radians(latitude)
        lon = math.radians(longitude)

        approx_earth_radius = 6371
        hypotenuse_distance = math.sqrt(2 * (math.pow(half_side_in_km, 2)))

        lat_min = math.asin(
            math.sin(lat) * math.cos(hypotenuse_distance / approx_earth_radius)
            + math.cos(lat)
            * math.sin(hypotenuse_distance / approx_earth_radius)
            * math.cos(225 * (math.pi / 180)),
        )
        lon_min = lon + math.atan2(
            math.sin(225 * (math.pi / 180))
            * math.sin(hypotenuse_distance / approx_earth_radius)
            * math.cos(lat),
            math.cos(hypotenuse_distance / approx_earth_radius)
            - math.sin(lat) * math.sin(lat_min),
        )

        lat_max = math.asin(
            math.sin(lat) * math.cos(hypotenuse_distance / approx_earth_radius)
            + math.cos(lat)
            * math.sin(hypotenuse_distance / approx_earth_radius)
            * math.cos(45 * (math.pi / 180)),
        )
        lon_max = lon + math.atan2(
            math.sin(45 * (math.pi / 180))
            * math.sin(hypotenuse_distance / approx_earth_radius)
            * math.cos(lat),
            math.cos(hypotenuse_distance / approx_earth_radius)
            - math.sin(lat) * math.sin(lat_max),
        )

        rad2deg = math.degrees

        return BoundingBox(
            min_latitude=rad2deg(lat_min),
            max_latitude=rad2deg(lat_max),
            min_longitude=rad2deg(lon_min),
            max_longitude=rad2deg(lon_max),
        )

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The OpenSky object.
        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
