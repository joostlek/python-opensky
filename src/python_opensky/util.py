"""Asynchronous Python client for OpenSky."""
from __future__ import annotations

from typing import Any, TypeVar

from .const import LOGGER

_EnumT = TypeVar("_EnumT")


def to_enum(
    enum_class: type[_EnumT],
    value: Any,
    default_value: _EnumT,
) -> _EnumT:
    """Convert a value to an enum and log if it doesn't exist."""
    try:
        return enum_class(value)  # type: ignore[call-arg]
    except ValueError:
        LOGGER.warning(
            "%s is an unsupported value for %s, please report this at https://github.com/joostlek/python-opensky/issues",
            value,
            str(enum_class),
        )
        return default_value
