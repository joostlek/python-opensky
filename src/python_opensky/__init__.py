"""Asynchronous Python client for the OpenSky API."""

from .const import AircraftCategory, PositionSource
from .exceptions import OpenSkyConnectionError, OpenSkyError
from .models import BoundingBox, StatesResponse, StateVector
from .opensky import OpenSky

__all__ = [
    "AircraftCategory",
    "BoundingBox",
    "OpenSky",
    "OpenSkyConnectionError",
    "OpenSkyError",
    "PositionSource",
    "StateVector",
    "StatesResponse",
]
