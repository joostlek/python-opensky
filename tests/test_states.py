"""Tests for the OpenSky Library."""
import asyncio
from dataclasses import asdict

import aiohttp
import pytest
from aiohttp import BasicAuth, ClientError
from aiohttp.web_request import BaseRequest
from aresponses import Response, ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from python_opensky import (
    BoundingBox,
    OpenSky,
    OpenSkyConnectionError,
    OpenSkyError,
    StatesResponse,
)
from python_opensky.exceptions import OpenSkyUnauthenticatedError

from . import load_fixture

OPENSKY_URL = "opensky-network.org"


async def test_states(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
) -> None:
    """Test retrieving states."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        response: StatesResponse = await opensky.get_states()
        assert asdict(response) == snapshot
        await opensky.close()


async def test_unavailable_states(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving no states."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("unavailable_states.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        response: StatesResponse = await opensky.get_states()
        assert response.states is not None
        assert len(response.states) == 0
        assert response.time == 1683488744
        await opensky.close()


async def test_own_states(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving own states."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    aresponses.add(
        OPENSKY_URL,
        "/api/states/own",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        await opensky.authenticate(
            BasicAuth(login="test", password="test"),
            contributing_user=True,
        )
        response: StatesResponse = await opensky.get_own_states()
        assert len(response.states) == 4
        assert opensky.opensky_credits == 8000
        assert opensky.remaining_credits() == 7999
        await opensky.close()


async def test_unavailable_own_states(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving no own states."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    aresponses.add(
        OPENSKY_URL,
        "/api/states/own",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("unavailable_states.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        await opensky.authenticate(
            BasicAuth(login="test", password="test"),
            contributing_user=True,
        )
        response: StatesResponse = await opensky.get_own_states()
        assert response.states is not None
        assert len(response.states) == 0
        assert response.time == 1683488744
        await opensky.close()


async def test_states_with_bounding_box(
    aresponses: ResponsesMockServer,
) -> None:
    """Test retrieving states."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all?time=0&extended=true&lamin=0&lamax=0&lomin=0&lomax=0",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
        match_querystring=True,
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        bounding_box = BoundingBox(
            min_latitude=0,
            max_latitude=0,
            min_longitude=0,
            max_longitude=0,
        )
        await opensky.get_states(bounding_box=bounding_box)
        await opensky.close()


async def test_credit_usage(
    aresponses: ResponsesMockServer,
) -> None:
    """Test credit usage."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        await opensky.get_states()
        assert opensky.remaining_credits() == opensky.opensky_credits - 4
        await opensky.close()


async def test_new_session(
    aresponses: ResponsesMockServer,
) -> None:
    """Test that it creates a new session if not given one."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )
    async with OpenSky() as opensky:
        assert not opensky.session
        await opensky.get_states()
        assert opensky.session


async def test_timeout(aresponses: ResponsesMockServer) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: BaseRequest) -> Response:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return aresponses.Response(body="Goodmorning!")

    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        response_handler,
    )

    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session, request_timeout=1)
        with pytest.raises(OpenSkyConnectionError):
            assert await opensky.get_states()
        await opensky.close()


async def test_auth(aresponses: ResponsesMockServer) -> None:
    """Test request authentication."""

    def response_handler(request: BaseRequest) -> Response:
        """Response handler for this test."""
        assert request.headers
        assert request.headers["Authorization"]
        assert request.headers["Authorization"] == "Basic dGVzdDp0ZXN0"
        return aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        )

    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        response_handler,
        repeat=2,
    )

    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        await opensky.authenticate(BasicAuth(login="test", password="test"))
        await opensky.get_states()
        await opensky.close()


async def test_unauthorized(aresponses: ResponsesMockServer) -> None:
    """Test request authentication."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=401,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
    )

    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        try:
            await opensky.authenticate(BasicAuth(login="test", password="test"))
            pytest.fail("Should've thrown exception")
        except OpenSkyUnauthenticatedError:
            pass
        assert opensky.is_authenticated is False
        await opensky.close()


async def test_user_credits(aresponses: ResponsesMockServer) -> None:
    """Test authenticated user credits."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixture("states.json"),
        ),
        repeat=2,
    )
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        assert opensky.opensky_credits == 400
        await opensky.authenticate(BasicAuth(login="test", password="test"))
        assert opensky.opensky_credits == 4000
        await opensky.authenticate(
            BasicAuth(login="test", password="test"),
            contributing_user=True,
        )
        assert opensky.opensky_credits == 8000
        assert opensky.is_contributing_user is True
        await opensky.close()


async def test_request_error(aresponses: ResponsesMockServer) -> None:
    """Test request error."""

    async def response_handler(_: BaseRequest) -> Response:
        """Response handler for this test."""
        raise ClientError

    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        response_handler,
    )

    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        with pytest.raises(OpenSkyConnectionError):
            assert await opensky.get_states()
        await opensky.close()


async def test_unexpected_server_response(
    aresponses: ResponsesMockServer,
) -> None:
    """Test handling a server error."""
    aresponses.add(
        OPENSKY_URL,
        "/api/states/all",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "plain/text"},
            text="Yes",
        ),
    )

    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        with pytest.raises(OpenSkyError):
            assert await opensky.get_states()
        await opensky.close()


async def test_unauthenticated_own_states() -> None:
    """Test unauthenticated access to own states."""
    async with aiohttp.ClientSession() as session:
        opensky = OpenSky(session=session)
        with pytest.raises(OpenSkyUnauthenticatedError):
            assert await opensky.get_own_states()
        await opensky.close()


async def test_calculating_credit_usage() -> None:
    """Test calculating credit usage."""
    opensky = OpenSky()
    bounding_box = BoundingBox(
        min_latitude=49.7,
        max_latitude=50.5,
        min_longitude=3.2,
        max_longitude=4.6,
    )
    assert opensky.calculate_credit_costs(bounding_box) == 1
    bounding_box = BoundingBox(
        min_latitude=46.5,
        max_latitude=49.9,
        min_longitude=-1.4,
        max_longitude=6.8,
    )
    assert opensky.calculate_credit_costs(bounding_box) == 2
    bounding_box = BoundingBox(
        min_latitude=42.2,
        max_latitude=49.8,
        min_longitude=-4.7,
        max_longitude=10.9,
    )
    assert opensky.calculate_credit_costs(bounding_box) == 3
    bounding_box = BoundingBox(
        min_latitude=42.2,
        max_latitude=49.8,
        min_longitude=-80.7,
        max_longitude=10.9,
    )
    assert opensky.calculate_credit_costs(bounding_box) == 4
