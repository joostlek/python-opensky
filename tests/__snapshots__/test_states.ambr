# serializer version: 1
# name: test_states
  dict({
    'states': list([
      dict({
        'barometric_altitude': 2217.42,
        'callsign': 'UAL421  ',
        'category': 4,
        'geo_altitude': 2194.56,
        'icao24': 'ab1644',
        'last_contact': 1683488743,
        'latitude': 42.5372,
        'longitude': -71.1656,
        'on_ground': False,
        'origin_country': 'United States',
        'position_source': 0,
        'sensors': None,
        'special_purpose_indicator': False,
        'time_position': 1683488743,
        'transponder_code': None,
        'true_track': 342.17,
        'velocity': 137.8,
        'vertical_rate': 13,
      }),
      dict({
        'barometric_altitude': None,
        'callsign': 'LPE2264 ',
        'category': 1,
        'geo_altitude': None,
        'icao24': 'e8027e',
        'last_contact': 1683488722,
        'latitude': -12.0115,
        'longitude': -77.1296,
        'on_ground': True,
        'origin_country': 'Chile',
        'position_source': 0,
        'sensors': None,
        'special_purpose_indicator': False,
        'time_position': 1683488479,
        'transponder_code': None,
        'true_track': 149.16,
        'velocity': 83.29,
        'vertical_rate': 9.1,
      }),
      dict({
        'barometric_altitude': 4335.78,
        'callsign': 'LPE2392 ',
        'category': 1,
        'geo_altitude': None,
        'icao24': 'e8027d',
        'last_contact': 1683488743,
        'latitude': -11.905,
        'longitude': -76.9275,
        'on_ground': False,
        'origin_country': 'Chile',
        'position_source': 0,
        'sensors': None,
        'special_purpose_indicator': False,
        'time_position': 1683488743,
        'transponder_code': None,
        'true_track': 332.17,
        'velocity': 187.32,
        'vertical_rate': 10.4,
      }),
      dict({
        'barometric_altitude': 701.04,
        'callsign': 'N28CN   ',
        'category': 9,
        'geo_altitude': 685.8,
        'icao24': 'a2cbe1',
        'last_contact': 1683488743,
        'latitude': 44.9762,
        'longitude': -93.9953,
        'on_ground': False,
        'origin_country': 'United States',
        'position_source': 0,
        'sensors': None,
        'special_purpose_indicator': False,
        'time_position': 1683488743,
        'transponder_code': None,
        'true_track': 293.15,
        'velocity': 73.29,
        'vertical_rate': -4.55,
      }),
    ]),
    'time': 1683488744,
  })
# ---
